From 86e0cac8495638432ca066afd5b550aa32368047 Mon Sep 17 00:00:00 2001
From: Joshua Strobl <joshua@stroblindustries.com>
Date: Fri, 11 Jan 2019 23:09:39 +0200
Subject: [PATCH 1/1] Fall back to org.gnome.desktop.interface if wingpanel
 datetime schema does not exist.

This commit will the case where the io.elementary.desktop.wingpanel.datetime schema doesn't exist, which would otherwise be fatal on platforms which support / ship applications which use Granite.DateTime.is_clock_format_12h() but do not support / ship Wingpanel OOTB.
---
 lib/DateTime.vala | 20 +++++++++++++++++---
 1 file changed, 17 insertions(+), 3 deletions(-)

diff --git a/lib/DateTime.vala b/lib/DateTime.vala
index 2069e1f..a9b1647 100644
--- a/lib/DateTime.vala
+++ b/lib/DateTime.vala
@@ -22,6 +22,8 @@
  * getting the default translated format for either date and time.
  */
 namespace Granite.DateTime {
+    private const string WINGPANEL_DATETIME_SCHEMA = "io.elementary.desktop.wingpanel.datetime";
+
     /**
      * Gets a default translated time format.
      * The function constructs a new string interpreting the //is_12h// and //with_second// parameters
@@ -110,9 +112,21 @@ namespace Granite.DateTime {
      * @return true if the clock format is 12h based, false otherwise.
      */
     private static bool is_clock_format_12h () {
-        var h24_settings = new Settings ("io.elementary.desktop.wingpanel.datetime");
-        var format = h24_settings.get_string ("clock-format");
-        return (format.contains ("12h"));
+        SettingsSchemaSource? source = SettingsSchemaSource.get_default();
+        bool use_wingpanel_schema = false;
+
+        if (source != null) { // If we have a valid SettingsSchemaSource
+                use_wingpanel_schema = (source.lookup(WINGPANEL_DATETIME_SCHEMA, true) != null);
+        }
+
+        Settings h24_settings = new Settings(use_wingpanel_schema ? WINGPANEL_DATETIME_SCHEMA : "org.gnome.desktop.interface");
+                
+        if (h24_settings != null) { // Neither schema exists
+                var format = h24_settings.get_string ("clock-format");
+                return format.contains("12h");
+        } else {
+                return false;
+        }
     }
 
     /**
-- 
2.20.1

